core/src/com/google/inject/internal/InternalMethodHandles.java

```diff
@@ -865,436 +879,118 @@ public String toString() {
-      {
-        // generate a default constructor
-        // This is just a default constructor that calls the super class constructor.
-        // basically `super(injector, factory, dependency?);`
-        int stackHeight = 0;
-        MethodVisitor mv =
-            cw.visitMethod(
-                ACC_PUBLIC,
-                "<init>",
-                ctorType.toMethodDescriptorString(),
-                /* signature= */ null,
-                /* exceptions= */ null);
-        mv.visitCode();
-        mv.visitVarInsn(ALOAD, stackHeight++); // this
-        mv.visitVarInsn(ALOAD, stackHeight++); // injector
-        mv.visitVarInsn(ALOAD, stackHeight++); // factory
-        if (!isScoped) {
-          mv.visitVarInsn(ALOAD, stackHeight++); // maybe dependency
+    private MethodHandle getHandle() {
+      var local = this.handle;
+      if (local == null) {
+        // synchronize on the factory instead of `this` since users might be using the provider
+        // itself as a lock.
+        synchronized (factory) {
+          local = this.handle;
+          if (local == null) {
+            // Since the Dependency is a constant for all calls to this handle, insert it so the jit
+            // can see it.
+            local =
+                MethodHandles.insertArguments(
+                    factory.getHandle(new LinkageContext(), /* linked= */ false), 1, dependency);
+            this.handle = local;
+          }
         }
-        mv.visitMethodInsn(
-            INVOKESPECIAL,
-            baseType.getInternalName(),
-            "<init>",
-            ctorType.toMethodDescriptorString(),
-            /* isInterface= */ false);
-        mv.visitInsn(RETURN);
-        mv.visitMaxs(
-            /* maxStack= */ stackHeight, // Just pushing the parameters
-            /* maxLocals= */ stackHeight // all the parameters
-            );
-        mv.visitEnd();
-      }
-      {
-        String descriptor = FACTORY_TYPE.toMethodDescriptorString();
-        // generate a doGet() method
-        MethodVisitor mv =
-            cw.visitMethod(
-                ACC_PROTECTED | ACC_FINAL,
-                "doGet",
-                descriptor,
-                /* signature= */ null,
-                /* exceptions= */ null);
-        mv.visitCode();
-        mv.visitVarInsn(ALOAD, 1); // ctx
-        mv.visitInvokeDynamicInsn("get", descriptor, BOOSTRAP_HANDLE);
-        mv.visitInsn(ARETURN);
-        mv.visitMaxs(
-            /* maxStack= */ 1, // Just the return value of the dynamic call
-            /* maxLocals= */ 2 // Just the 'this' and ctx parameters
-            );
-        mv.visitEnd();
-      }
-      // We must define the class as collectable so that it can be garbage collected.  The static
-      // field will hold a reference to an InternalFactory which may hold references to the
-      // injector.
-      Class<?> clazz;
-      try {
-        clazz =
-            ClassDefining.defineCollectable(
```

Comment:
Should we also delete all the ClassDefining stuff introduced in the earlier changes ?  AFAICT we don't need "collectable" infra anymore.

Reply:
Yes. Will delete!